crates/gitbutler-repo/src/change_reference.rs

```diff
@@ -0,0 +1,213 @@
+use crate::{LogUntil, RepoActionsExt};
+use anyhow::Context;
+use anyhow::{anyhow, Result};
+use gitbutler_branch::ChangeReference;
+use gitbutler_branch::VirtualBranchesHandle;
+use gitbutler_branch::{Branch, BranchId};
+use gitbutler_command_context::CommandContext;
+use gitbutler_commit::commit_ext::CommitExt;
+use gitbutler_reference::ReferenceName;
+use itertools::Itertools;
+
+/// Given a branch id, returns the the GitButler references associated with the branch.
+/// References within the same branch effectively represent a stack of sub-branches.
+pub fn list_branch_references(
+    ctx: &CommandContext,
+    branch_id: BranchId,
+) -> Result<Vec<ChangeReference>> {
+    let handle = VirtualBranchesHandle::new(ctx.project().gb_dir());
+    let vbranch = handle.get_branch(branch_id)?;
+    Ok(vbranch.references)
+}
+
+/// Creates a new virtual branch reference and associates it with the branch.
+/// However this will return an error if:
+///   - a reference for the same commit already exists, an error is returned.
+///   - the reference name already exists, an error is returned.
+pub fn create_change_reference(
+    ctx: &CommandContext,
+    branch_id: BranchId,
+    name: ReferenceName,
+    change_id: String,
+) -> Result<ChangeReference> {
+    // The reference must be parseable as a remote reference
+    gitbutler_reference::RemoteRefname::from_str(&name)
+        .context("Failed to parse the provided reference")?;
+    let handle = VirtualBranchesHandle::new(ctx.project().gb_dir());
+
+    // The branch must exist
+    let mut vbranch = handle.get_branch(branch_id)?;
+
+    // Enusre that the commit acutally exists
+    // let commit = ctx
+    //     .repository()
+    //     .find_commit(commit_id)
+    //     .context(anyhow!("Commit {} does not exist", commit_id))?;
+    let commit = commit_by_branch_id_and_change_id(ctx, &vbranch, &handle, change_id)?;
+
+    let change_id = commit
+        .change_id()
+        .ok_or(anyhow!("Commit {} does not have a change id", commit.id()))?;
+
+    let branch_reference = ChangeReference {
+        name,
+        branch_id,
+        change_id: change_id.clone(),
+    };
+    let all_references = handle
+        .list_all_branches()?
+        .into_iter()
+        .flat_map(|branch| branch.references)
+        .collect_vec();
+    // Ensure the reference name does not already exist
+    if all_references
+        .iter()
+        .any(|r| r.name == branch_reference.name)
+    {
+        return Err(anyhow!(
+            "A reference {} already exists",
+            branch_reference.name
+        ));
+    }
+    // Ensure the change is not already referenced
+    if all_references.iter().any(|r| r.change_id == change_id) {
+        return Err(anyhow!(
+            "A reference for change {} already exists",
+            change_id
+        ));
+    }
+    validate_commit(&vbranch, commit.id(), ctx, &handle)?;
+    vbranch.references.push(branch_reference.clone());
```

Comment:
Interesting, it looks like these references are fully 'virtual', no Git reference exists for it.
Doing that certainly has the advantage that the user can't alter them, nor will they see them outside of GB.

Reply:
yep - these are also references to a change id which provides some stability (eg. when a message is reworded etc)